overlord/snapstate/backend/snapdata.go

```diff
@@ -68,19 +68,27 @@ func (b Backend) RemoveSnapSaveData(snapInfo *snap.Info, dev snap.Device) error
 	return os.RemoveAll(saveDir)
 }
 
-// RemoveSnapDataDir removes base snap data directory
-func (b Backend) RemoveSnapDataDir(info *snap.Info, hasOtherInstances bool) error {
+// RemoveSnapDataDir removes base snap data directories
+func (b Backend) RemoveSnapDataDir(info *snap.Info, hasOtherInstances bool, opts *dirs.SnapDirOptions) error {
 	if info.InstanceKey != "" {
 		// data directories of snaps with instance key are never used by
 		// other instances
-		if err := os.Remove(snap.BaseDataDir(info.InstanceName())); err != nil && !os.IsNotExist(err) {
+		dirs, err := snapBaseDataDirs(info, opts)
+		if err != nil {
+			return err
+		}
+		if err = removeDirs(dirs); err != nil {
 			return fmt.Errorf("failed to remove snap %q base directory: %v", info.InstanceName(), err)
 		}
 	}
 	if !hasOtherInstances {
 		// remove the snap base directory only if there are no other
 		// snap instances using it
-		if err := os.Remove(snap.BaseDataDir(info.SnapName())); err != nil && !os.IsNotExist(err) {
+		dirs, err := snapBaseDataDirs(info, opts)
+		if err != nil {
+			return err
+		}
+		if err = removeDirs(dirs); err != nil {
```

Comment:
If I follow things correctly, I think what we're removing should always be empty directories? It looks like `removeDirs` uses `os.RemoveAll`, which would also remove files, if they're there. Could it use just `os.Remove` to prevent us from removing anything we don't want to?

Reply:
What happens if a user creates a symlink from `~/snap/<snap-name>/current` to `/` - will snapd start `rm -rf ...` everything?

Reply:
Great point, updated to use `os.Remove()` on data directory and broken symlink only. Thank!